### DISCUSSION TITLE
Code-Style - Forward Declarations

### DISCUSSION THREAD

**Discussion by kykrueger** [nebulastream/nebulastream#19] [Polls]
There seems to be some contention in the c++ community about the usage of forward declarations. 
The google-style-guide for example advises to instead include the headers you need instead of forward declarations where possible.

> If a source or header file refers to a symbol defined elsewhere, the file should directly include a header file which properly intends to provide a declaration or definition of that symbol. It should not include header files for any other reason.
> 
> Do not rely on transitive inclusions. This allows people to remove no-longer-needed #include statements from their headers without breaking clients. This also applies to related headers - foo.cc should include bar.h if it uses a symbol from it even if foo.h includes bar.h.

The usage of forward declarations has pros and cons. Here is a list combining pros and cons from a few sources. 

Pros:
* Forward declarations can save compile time, as #includes force the compiler to open more files and process more input.
* Forward declarations can save on unnecessary recompilation. #includes can force your code to be recompiled more often, due to unrelated changes in the header.
* Minimize dependencies.

Cons:

* Forward declarations can hide a dependency, allowing user code to skip necessary recompilation when headers change.
* A forward declaration as opposed to an #include statement makes it difficult for automatic tooling to discover the module defining the symbol.
* A forward declaration may be broken by subsequent changes to the library. Forward declarations of functions and templates can prevent the header owners from making otherwise-compatible changes to their APIs, such as widening a parameter type, adding a template parameter with a default value, or migrating to a new namespace.
* Forward declaring symbols from namespace std:: yields undefined behavior.
* It can be difficult to determine whether a forward declaration or a full #include is needed. Replacing an #include with a forward declaration can silently change the meaning of code:

```
// b.h:
struct B {};
struct D : B {};

// good_user.cc:
#include "b.h"
void f(B*);
void f(void*);
void test(D* x) { f(x); }  // Calls f(B*)
```

* If the #include was replaced with forward decls for B and D, test() would call f(void*).
* Forward declaring multiple symbols from a header can be more verbose than simply #includeing the header.
* Structuring code to enable forward declarations (e.g., using pointer members instead of object members) can make the code slower and more complex.

For more information to aid the discussion, see this issue by @varpande 
https://github.com/nebulastream/nebulastream/issues/4062

**Comment by alepping** [comment]
I'm very torn on this topic, so I won't vote. However, I fully agree that we should not randomly mix forward declarations and includes.
I also like the idea of always importing everything required for robustness and to **not** rely on transitive includes.
If it is possible to avoid situations like the TestTupleBuffer (former DynamicTupleBuffer), where a single change in the header file caused 300+ files to recompile, by using includes more awarely, then I fully support avoiding forward declarations.

**Comment by alepping** [comment]
Additionally, Forward Declarations allow to avoid dependencies in header files.
Class A might depend on class B and vice versa, if one of the classes forward declares the other class in the header and only imports it in the implementation. I'm unsure whether it is strictly bad design to make use of this.
